core/trie/src/node_codec.rs

```diff
@@ -18,72 +18,93 @@
 
 use rstd::marker::PhantomData;
 use rstd::vec::Vec;
+use rstd::borrow::Borrow;
 use codec::{Encode, Decode, Compact};
 use hash_db::Hasher;
-use trie_db::{self, DBValue, NibbleSlice, node::Node, ChildReference};
+use trie_db::{self, NibbleSlice, node::Node, ChildReference, BitMap,
+	NibbleOps, Partial, NodeCodec as NodeCodecT, ChildSliceIx};
 use crate::error::Error;
-use super::{EMPTY_TRIE, LEAF_NODE_OFFSET, LEAF_NODE_BIG, EXTENSION_NODE_OFFSET,
-	EXTENSION_NODE_BIG, take, partial_to_key, node_header::NodeHeader, branch_node};
+use crate::trie_constants;
+use super::{node_header::{NodeHeader, NodeKind}};
+
+fn take<'a>(input: &mut &'a[u8], count: usize) -> Option<&'a[u8]> {
+	if input.len() < count {
+		return None
+	}
+	let r = &(*input)[..count];
+	*input = &(*input)[count..];
+	Some(r)
+}
 
 /// Concrete implementation of a `NodeCodec` with Parity Codec encoding, generic over the `Hasher`
 #[derive(Default, Clone)]
-pub struct NodeCodec<H: Hasher>(PhantomData<H>);
+pub struct NodeCodec<H, N, BM>(PhantomData<(H, N, BM)>);
 
-impl<H: Hasher> trie_db::NodeCodec<H> for NodeCodec<H> {
+impl<
+	H: Hasher,
+	N: NibbleOps,
+	BM: BitMap<Error = Error>
+> NodeCodecT<H, N> for NodeCodec<H, N, BM> {
 	type Error = Error;
 
-	fn hashed_null_node() -> H::Out {
-		H::hash(&[0u8][..])
+	fn hashed_null_node() -> <H as Hasher>::Out {
+		H::hash(<Self as NodeCodecT<_, N>>::empty_node())
 	}
 
-	fn decode(data: &[u8]) -> ::rstd::result::Result<Node, Self::Error> {
-		use Error::BadFormat;
+	fn decode(data: &[u8]) -> rstd::result::Result<Node<N>, Self::Error> {
 		let input = &mut &*data;
-		match NodeHeader::decode(input).ok_or(BadFormat)? {
+		let head = NodeHeader::decode(input).ok_or(Error::BadFormat)?;
+		match head {
 			NodeHeader::Null => Ok(Node::Empty),
-			NodeHeader::Branch(has_value) => {
-				let bitmap = u16::decode(input).ok_or(BadFormat)?;
+			NodeHeader::Branch(has_value, nibble_count) => {
+				let nb_nibble_hpe = nibble_count % N::NIBBLE_PER_BYTE;
+				if nb_nibble_hpe > 0
+					&& N::masked_left((N::NIBBLE_PER_BYTE - nb_nibble_hpe) as u8, input[0]) != 0 {
+					return Err(Error::BadFormat);
+				}
+				let nibble_data = take(input, (nibble_count + (N::NIBBLE_PER_BYTE - 1)) / N::NIBBLE_PER_BYTE)
```

Comment:
`N::` - non-local identifiers should generally be longer than two letters and certainly be complete words.